### ISSUE TITLE
Performance Improvement

### ISSUE THREAD

**Issue by yancyribbens** [p2pderivatives/rust-bitcoin-coin-selection#5]
I made a rather serious error which resulted in poor performance in the initial implementation.  I've made a fork of the project and added a benchmark to [measure performance](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/120833c1530680a20abd894f42b7ea5b0496b6b3).  By removing the redundant for loop [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/120833c1530680a20abd894f42b7ea5b0496b6b3) the time to search through a utxo space of `10,000` with `cargo bench`:

`time:   [315.80 ms 316.59 ms 317.32 ms]`

to the improved:

`time:   [438.80 µs 441.28 µs 443.84 µs]`
`change: [-99.861% -99.860% -99.859%] (p = 0.00 < 0.05)`

The search performance is now measured in nano-seconds instead of micro-seconds, an increase of `1000x`.

My apologies for the initial oversight.

**Comment by Xekyo** [comment]
Nice, that's a solid improvement

**Comment by yancyribbens** [comment]
fyi: I made some minor improvements to the readability of theses commits ([here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/75d80040095a9f58ed000e1626de532e3776f91a) and [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/69ec453aa9752e0746a1b228ae224924b7f1888a).

**Comment by yancyribbens** [comment]
I've added a few more performance improvements, most notably this one [here:](https://github.com/yancyribbens/rust-bitcoin-coin-selection/commit/bae7bab67c3e90ca8d4db510f0296b95fdf1242b).  I think the commit simplifies the code as well as a `20%` reduction in run-time.  Overall, I've managed to reduce the time down to ~250 nano-seconds for a search space of 10,000 utxos:

`time:   [254.42 µs 254.53 µs 254.64 µs]`
`change: [-22.062% -21.645% -21.349%] (p = 0.00 < 0.05)`

**Comment by yancyribbens** [comment]
@Xekyo thanks for taking a look at my previous improvement btw.  Please let me know if you have any suggestions.

**Comment by yancyribbens** [comment]
I just noticed that while the [commit](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/75d80040095a9f58ed000e1626de532e3776f91a) improves performance, it actually precludes more optimal solutions, and so while I thought I had made a mistake, the second for loop is actually required.

The test [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/commit/01f59134c70522b3bf45e0bf87486cf0a3645b38) fails on the "performance improved" version however the original solution is correct.

Long story short, false alarm.
